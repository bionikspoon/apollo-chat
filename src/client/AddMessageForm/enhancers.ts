import gql from 'graphql-tag'
import {
  ChildDataProps,
  ChildMutateProps,
  compose,
  graphql,
} from 'react-apollo'
import { RouteComponentProps, withRouter } from 'react-router'

interface IData {
  username: string
}
interface IResponse {
  id: string
  user: string
  body: string
  date: string
}

const GET_USERNAME = gql`
  query GetUsername {
    username @client
  }
`

const ADD_MESSAGE_MUTATION = gql`
  mutation AddMessageForm($user: String!, $body: String!) {
    addMessage(user: $user, body: $body) {
      id
      user
      body
      date
    }
  }
`

export default compose(
  withRouter,
  graphql(GET_USERNAME),
  graphql(ADD_MESSAGE_MUTATION, { name: 'addMessage' })
)

export type AddMessageFormProps = ChildMutateProps<
  ChildDataProps<RouteComponentProps, IData>,
  IResponse
>
