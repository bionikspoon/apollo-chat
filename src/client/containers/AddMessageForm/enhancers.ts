import { Intent } from '@blueprintjs/core'
import gql from 'graphql-tag'
import { ChildDataProps, compose, graphql, MutationFunc } from 'react-apollo'
import { RouteComponentProps, withRouter } from 'react-router'
import { IUser } from '../../setupApollo/clientState'

type IData = IUser
interface IAddMessageResponse {
  id: string
  user: string
  body: string
  date: string
}

interface IAddMessageVariables {
  body: string
  username: string
  color: Intent
}

const GET_USERNAME = gql`
  query GetUsername {
    user @client {
      username
      color
    }
  }
`

const ADD_MESSAGE_MUTATION = gql`
  mutation AddMessageForm($username: String!, $body: String!, $color: String!) {
    addMessage(user: $username, body: $body, color: $color) {
      id
      user
      body
      date
      color
    }
  }
`

export default compose(
  withRouter,
  graphql(GET_USERNAME),
  graphql(ADD_MESSAGE_MUTATION, { name: 'addMessage' })
)

export type AddMessageFormProps = ChildDataProps<RouteComponentProps, IData> & {
  addMessage: MutationFunc<IAddMessageResponse, IAddMessageVariables>
}
