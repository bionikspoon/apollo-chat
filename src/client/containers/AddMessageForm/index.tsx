import AddMessageForm from './AddMessageForm'
import enhancer from './enhancers'

export default enhancer(AddMessageForm)
