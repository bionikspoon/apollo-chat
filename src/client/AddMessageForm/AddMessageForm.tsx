import { Button, ControlGroup, InputGroup, Tag } from '@blueprintjs/core'
import { css, StyleSheet } from 'aphrodite'
import * as React from 'react'
import { ChangeEvent, Component, FormEvent } from 'react'

import { AddMessageFormProps } from './enhancers'

export default class AddMessageForm extends Component<AddMessageFormProps, {}> {
  public state = { body: '' }

  public render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <ControlGroup fill={true}>
          <Tag intent="success" className={css(styles.tag)}>
            Manu
          </Tag>
          <InputGroup
            type="text"
            value={this.state.body}
            onChange={this.handleInputChange}
            rightElement={
              <Button type="submit" icon="arrow-right" intent="success" />
            }
          />
        </ControlGroup>
      </form>
    )
  }

  private handleSubmit = (event: FormEvent) => {
    event.preventDefault()

    this.props.mutate({
      variables: {
        body: this.state.body,
        user: 'Manu',
      },
    })

    this.setState({ body: '' })
  }

  private handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    this.setState({ body: event.target.value })
  }
}

const styles = StyleSheet.create({
  tag: { flexGrow: 0 },
})
