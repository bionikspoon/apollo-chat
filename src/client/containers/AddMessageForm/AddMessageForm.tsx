import { Button, ControlGroup, InputGroup, Tag } from '@blueprintjs/core'
import { css, StyleSheet } from 'aphrodite'
import * as React from 'react'
import { ChangeEvent, Component, FormEvent } from 'react'
import { AddMessageFormProps } from './enhancers'

export default class AddMessageForm extends Component<AddMessageFormProps, {}> {
  public state = { body: '' }

  public componentDidMount() {
    if (!this.props.data.user) return
    const { username } = this.props.data.user
    if (username.length) return

    this.props.history.replace('/')
  }

  public render() {
    if (!this.props.data.user) return
    const { color, username } = this.props.data.user

    return (
      <form onSubmit={this.handleSubmit}>
        <ControlGroup fill={true}>
          <Tag intent={color} className={css(styles.tag)}>
            {username}
          </Tag>
          <InputGroup
            type="text"
            value={this.state.body}
            onChange={this.handleInputChange}
            intent={color}
            rightElement={
              <Button type="submit" icon="arrow-right" intent={color} />
            }
          />
        </ControlGroup>
      </form>
    )
  }

  private handleSubmit = (event: FormEvent) => {
    event.preventDefault()

    if (!this.props.data.user) return
    const { color, username } = this.props.data.user
    const { body } = this.state

    if (!this.state.body.length) return

    this.props.addMessage({
      variables: {
        body,
        color,
        username,
      },
    })

    this.setState({ body: '' })
  }

  private handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    this.setState({ body: event.target.value })
  }
}

const styles = StyleSheet.create({
  tag: { flexGrow: 0 },
})
