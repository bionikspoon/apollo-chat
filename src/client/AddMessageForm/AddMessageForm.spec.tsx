import { shallow } from 'enzyme'
import * as React from 'react'

import AddMessageForm from '.'

it('matches snapshot', () => {
  const wrapper = shallow(<AddMessageForm />)
  expect(wrapper).toMatchSnapshot()
})
