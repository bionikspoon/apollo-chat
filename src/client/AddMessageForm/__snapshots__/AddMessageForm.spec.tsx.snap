// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Mutation
  ignoreResults={true}
  mutation={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "AddMessageForm",
          },
          "operation": "mutation",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "user",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "user",
                      },
                    },
                  },
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "body",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "body",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "addMessage",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "user",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "body",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "date",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "user",
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "body",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 156,
        "start": 0,
      },
    }
  }
>
  <Component />
</Mutation>
`;
